_posts/2016-06-30-state-in-scalable-architectures.md

```diff
@@ -0,0 +1,173 @@
+
+>> One final point is that the type of complexity we are discussing in this paper is that which makes large systems hard to understand. It is this that causes us to expend huge resources in creating and maintaining such systems. This type of complexity has nothing to do with complexity theory — the branch of computer science which studies the resources consumed by a machine executing a program.
+
+## State at service level
+
+Through this post we'll see how simplicity and scalability aligns blissfully when handling state. Let's start with approaches to handle state at service level. We'll move then into distributed systems integration.
+
+### Global State
+
+This is the loosest approach, e.g. `public static` non final variables in Java. How can we informal reason about code that uses global variables? We need to inspect all the code that is mutating that variable, and those lines of code might live in completely unrelated parts of the system. Having to reason about plenty of pieces at the same time is inefficient and dangerous. Global state forces us to do black box testing as the pieces of our system are coupled through global state.
+
+A priori simple lines of code get quickly 'contaminated' by the complexity derived of state:
+
+>> The key problem is that a test (of any kind) on a system or component that is in one particular state tells you nothing at all about the behaviour of that system or component when it happens to be in another state.
+
+### Object Oriented Programming State
+
+>> In most forms of object-oriented programming (OOP) an object is seen as consisting of some state together with a set of procedures for accessing and manipulating that state.
+
+As [Uncle Bob](https://www.wikiwand.com/en/Robert_Cecil_Martin) said once, programming paradigms succeeds, paradoxically, by removing power. Java succeed because it removed from developers the power to manage memory. OOP strongly suggests that code that mutates state should be colocated with that state. That guarantee, along preconditions and postconditions, makes code much easier to reason about and test.
+
+Using inheritance to share code is a bad idea based on those premises. State mutators will be located in different files, even if they finally constitute a single object on runtime, obscuring the readability of the code.
+
+### Functional Programming State
+
+FP tries to solve the complexity problem avoiding mutable state. Thanks to that we can have functions whose outputs depend completely on their inputs. Avoiding is not as simple as closing your eyes, though. There are some scenarios where we need to handle state. Let's use an [example](https://github.com/fpinscala/fpinscala/blob/master/answers/src/main/scala/fpinscala/state/State.scala) from [Functional programming in Scala book](https://www.manning.com/books/functional-programming-in-scala):
```

Comment:
Nope. I've rephrased it: "FP tries to solve this complexity avoiding mutable state completely"